org.eclipse.winery.model.adaptation/src/test/java/org/eclipse/winery/model/adaptation/substitution/patterndetection/BehaviorPatternDetectionTest.java

```diff
@@ -0,0 +1,364 @@
+        topology.addNodeTemplate(candidateElement);
+        behaviorPatternDetection.applyRefinement(refinementCandidate, topology);
+        assertEquals(topology.getNodeTemplateOrRelationshipTemplate().size(), 1);
+        policies = topology.getNodeTemplates().get(0).getPolicies().getPolicy();
+        assertEquals(policies.size(), 2);
+        assertTrue(policies.stream().anyMatch(policy -> policy.getName().equals("noProp")));
+        assertTrue(policies.stream().anyMatch(policy -> policy.getName().equals("oneProp")));
+
+        // 'noProp' & 'oneProp' still applicable, 'multiProps' not applicable because not all properties are set
+        candidateProps.put("multiPropsProp1", "true");
+        topology.getNodeTemplateOrRelationshipTemplate().clear();
+        topology.addNodeTemplate(candidateElement);
+        behaviorPatternDetection.applyRefinement(refinementCandidate, topology);
+        assertEquals(topology.getNodeTemplateOrRelationshipTemplate().size(), 1);
+        policies = topology.getNodeTemplates().get(0).getPolicies().getPolicy();
+        assertEquals(policies.size(), 2);
+        assertTrue(policies.stream().anyMatch(policy -> policy.getName().equals("noProp")));
+        assertTrue(policies.stream().anyMatch(policy -> policy.getName().equals("oneProp")));
+
+        // 'noProp' & 'oneProp' still applicable, 'multiProps' not applicable because not all properties are set
+        candidateProps.put("multiPropsProp2", "true");
+        topology.getNodeTemplateOrRelationshipTemplate().clear();
+        topology.addNodeTemplate(candidateElement);
+        behaviorPatternDetection.applyRefinement(refinementCandidate, topology);
+        assertEquals(topology.getNodeTemplateOrRelationshipTemplate().size(), 1);
+        policies = topology.getNodeTemplates().get(0).getPolicies().getPolicy();
+        assertEquals(policies.size(), 2);
+        assertTrue(policies.stream().anyMatch(policy -> policy.getName().equals("noProp")));
+        assertTrue(policies.stream().anyMatch(policy -> policy.getName().equals("oneProp")));
+
+        // 'noProp' & 'oneProp' & 'multiProps' applicable
+        candidateProps.put("multiPropsProp3", "non-null and non-empty");
+        topology.getNodeTemplateOrRelationshipTemplate().clear();
+        topology.addNodeTemplate(candidateElement);
+        behaviorPatternDetection.applyRefinement(refinementCandidate, topology);
+        assertEquals(topology.getNodeTemplateOrRelationshipTemplate().size(), 1);
+        policies = topology.getNodeTemplates().get(0).getPolicies().getPolicy();
+        assertEquals(policies.size(), 3);
+        assertTrue(policies.stream().anyMatch(policy -> policy.getName().equals("noProp")));
+        assertTrue(policies.stream().anyMatch(policy -> policy.getName().equals("oneProp")));
+        assertTrue(policies.stream().anyMatch(policy -> policy.getName().equals("multiProps")));
+    }
+
+    private NamespaceManager namespaceManager() {
```

Comment:
Isn't there a NamespaceManager for testing already?

Reply:
Yes, I initially tried to use the `MockNamespaceManager` here but it's only accessible in the `topologygraph.matching` package, not in the `adaptation.substitution` package of this test. I could move the `MockNamespaceManager` to src/main of `topologygraph.matching` to make it accessible but I'm not sure if that's a better solution...?

Reply:
Okay, we use your version here.